### PR TITLE
Updated to Kafka 10

### PR DESCRIPTION
**Pull request by pablocasares**
The Kafka API has been updated to Kafka 10 and now the CEP supports keyed messages at the Kafka input/output. The CEP will produce the message to the output topic with the same key which they had at the input. If the message is consumed without key it will be produced to the partition 0. 

For example: 
- Input [key, message]: keyA, {"value":0,"type":"rx_pkts_tunnel"}
- Rule: 
  `{
  "id": "rule1",
  "input": ["inputTopic"],
  "output": { "outputtopic": "output_topic" },
  "executionPlan": "from inputTopic select value, type insert into outputtopic"
  }`

will produce as output:

`keyA, {"value":0,"type":"rx_pkts_tunnel"}`

and: 
- Input [key, message]:  null, {"value":0,"type":"rx_pkts_tunnel"}
- Same rule as above

will produce as output (at partition 0):
`null, {"value":0,"type":"rx_pkts_tunnel"}`

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -47,7 +47,7 @@
         <dependency>
```
Comment (andresgomezfrr) [inline pom.xml:47]: I think that you can use

``` xml
<dependency>
    <groupId>org.apache.kafka</groupId>
    <artifactId>kafka-clients</artifactId>
    <version>0.10.0.1</version>
</dependency>
```

File: src/main/java/net/redborder/cep/sinks/kafka/KafkaSink.java [hunk 2]
```diff
@@ -43,23 +47,13 @@ public KafkaSink(Map<String, Object> properties) {
...
-            props.put("metadata.broker.list", "127.0.0.1:9092");
-        }
-        props.put("serializer.class", "kafka.serializer.StringEncoder");
-        props.put("request.required.acks", "1");
-        props.put("message.send.max.retries", "60");
-        props.put("retry.backoff.ms", "1000");
-        props.put("producer.type", "async");
-        props.put("queue.buffering.max.messages", "10000");
-        props.put("queue.buffering.max.ms", "500");
+        props.put("bootstrap.servers", ConfigData.getKafkaBrokers());
```
Comment (andresgomezfrr) [inline src/main/java/net/redborder/cep/sinks/kafka/KafkaSink.java]: You can use the ProducerConfigs class to define the properties.

File: src/main/java/net/redborder/cep/sinks/kafka/KafkaSink.java [hunk 3]
```diff
@@ -86,22 +100,25 @@ public void process(String streamName, String topic, Map<String, Object> message
...
      */
 
     public void send(String topic, String key, Map<String, Object> message) {
         try {
             String messageStr = objectMapper.writeValueAsString(message);
-            KeyedMessage<String, String> keyedMessage = new KeyedMessage<>(topic, key, messageStr);
-            producer.send(keyedMessage);
+            ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, messageStr);
+            producer.send(record);
+            producer.flush();
```
Comment (andresgomezfrr) [inline src/main/java/net/redborder/cep/sinks/kafka/KafkaSink.java]: You don't need to do a flush each send. You need to do a flush when shutdown the system.

File: src/main/java/net/redborder/cep/sinks/kafka/SimplePartitioner.java [hunk 4]
```diff
@@ -10,19 +12,18 @@
...
-     * @return The index of the partition that will receive the message associated with the given key
-     */
-    public int partition(Object key, int a_numPartitions) {
-        String stringKey = (String) key;
-        int offset = stringKey.hashCode();
-        return Math.abs(offset % a_numPartitions);
+
+    @Override
+    public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
+        return (key == null) ? 0 : Math.abs(key.hashCode() % cluster.partitionCountForTopic(topic));
```
Comment (andresgomezfrr) [inline src/main/java/net/redborder/cep/sinks/kafka/SimplePartitioner.java]: Better thant hashCode you can use murmur2 

``` java
import org.apache.kafka.common.utils.Utils;

Utils.abs(Utils.murmur2(key.getBytes())) % numPartitions
```

File: src/main/java/net/redborder/cep/sinks/syslog/SyslogSink.java [hunk 5]
```diff
@@ -17,7 +16,18 @@ public SyslogSink(Map<String, Object> properties) {
 
     @Override
     public void process(String streamName, String topic, Map<String, Object> message) {
-        logger.info("RULE: [{}] ALERT: [{}]", streamName, message);
+        String key;
+        if (message.containsKey("__KEY")) {
```
Comment (andresgomezfrr) [inline src/main/java/net/redborder/cep/sinks/syslog/SyslogSink.java]: Create a utils class to define a string constant to __KEY

File: src/main/java/net/redborder/cep/sources/kafka/KafkaSource.java [hunk 6]
```diff
@@ -168,7 +152,10 @@ private void initConsumers() {
      */
 
     public void shutdown() {
-        if (consumer != null) consumer.shutdown();
+
+        for (Consumer consumer : consumers) {
```
Comment (andresgomezfrr) [inline src/main/java/net/redborder/cep/sources/kafka/KafkaSource.java]: I Think that you don't need this if you do 

``` java
executor.shutdown(); 
```

**Review by andresgomezfrr — CHANGES_REQUESTED** [review]
Add some integration test using EmbeddedSingleNodeKafkaCluster class.

**Comment by pablocasares** [comment]
The changes requested have been finished.

**Review by andresgomezfrr — APPROVED** [review]

**Pull request closed without merging**